crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -598,13 +601,36 @@ impl<'db> TypeInferenceBuilder<'db> {
         } = with_statement;
 
         for item in items {
-            self.infer_expression(&item.context_expr);
-            self.infer_optional_expression(item.optional_vars.as_deref());
+            if item.optional_vars.is_some() {
+                self.infer_definition(item);
+            } else {
+                self.infer_expression(&item.context_expr);
+            }
         }
 
         self.infer_body(body);
     }
 
+    fn infer_with_item_definition(
+        &mut self,
+        with_item: &ast::WithItem,
+        definition: Definition<'db>,
+    ) {
+        let ast::WithItem {
+            range: _,
+            context_expr,
+            optional_vars,
+        } = with_item;
+
+        let Some(optional_vars) = optional_vars.as_deref() else {
+            unreachable!("With item definition without optional vars");
+        };
```

Comment:
Nit: I prefer to use `expect`. It's shorter 
```suggestion
        let optional_vars = optional_vars.as_deref().expect("With item definition without optional vars");
        };
```